### ISSUE TITLE
Type error on SIMD native types not caught by compiler

### ISSUE THREAD

**Issue by jfalcou** [llvm/llvm-project#126491]
@SadiinsoSnowfall and I were cleaning up some more of our SIMD library code and found that code like: 

```c++
auto f(__m128d a, __m128 b) { return a + b; }
```

Compiles when it should fail.

Reproduction :  

https://godbolt.org/z/eE4b4Kdn8

**Comment by jfalcou** [comment]
It should just not compile no ?

**Comment by jfalcou** [comment]
OK, but as for Agner Fog, I can't really rely on people to use this. Is there any plan to fix it from within the intrinsic file ?

**Comment by RKSimon** [comment]
> OK, but as for Agner Fog, I can't really rely on people to use this. Is there any plan to fix it from within the intrinsic file ?

The intrinsics headers are tested with -flax-vector-conversions=none already so they should be clean, there's no intention to force apply compile flags based off including the intrinsics headers.

**Comment by jfalcou** [comment]
> > OK, but as for Agner Fog, I can't really rely on people to use this. Is there any plan to fix it from within the intrinsic file ?
> 
> The intrinsics headers are tested with -flax-vector-conversions=none already so they should be clean, there's no intention to force apply compile flags based off including the intrinsics headers.

clearly they are not then ?

**Comment by phoebewang** [comment]
> > > OK, but as for Agner Fog, I can't really rely on people to use this. Is there any plan to fix it from within the intrinsic file ?
> > 
> > 
> > The intrinsics headers are tested with -flax-vector-conversions=none already so they should be clean, there's no intention to force apply compile flags based off including the intrinsics headers.
> 
> clearly they are not then ?

```
$ grep -r flax-vector-conversions clang/test/CodeGen/X86/ | wc -l
199
```

**Comment by jfalcou** [comment]
Ok, so let me rephrase. Why should I use (or ask my user to use) a compiler flag to ensure the expected behaviour?

Or did I miss something from the discussion in #17538?

**Comment by jfalcou** [comment]
I beg to disagree. The behavior exposed in the first message is erroneous and should not work as expected without any specific options passed.

**Comment by jfalcou** [comment]
OK then. I guess PR are welcome on those remaining ones ?

**Comment by phoebewang** [comment]
> OK then. I guess PR are welcome on those remaining ones ?

Definitely
